domain/src/test/java/org/oppia/android/domain/oppialogger/logscheduler/MetricLogSchedulingWorkerTest.kt

```diff
@@ -0,0 +1,324 @@
+    ).build()
+
+    val request: OneTimeWorkRequest = OneTimeWorkRequestBuilder<LogUploadWorker>()
+      .setInputData(inputData)
+      .build()
+
+    workManager.enqueue(request)
+    testCoroutineDispatchers.runCurrent()
+    val workInfo = workManager.getWorkInfoById(request.id)
+
+    assertThat(workInfo.get().state).isEqualTo(WorkInfo.State.SUCCEEDED)
+    assertThat(
+      fakePerformanceMetricsEventLogger
+        .getMostRecentPerformanceMetricsEvent()
+        .loggableMetric
+        .loggableMetricTypeCase
+    ).isEqualTo(OppiaMetricLog.LoggableMetric.LoggableMetricTypeCase.NETWORK_USAGE_METRIC)
+  }
+
+  @Test
+  fun testWorker_enqueueRequest_verifyMemoryUsagePerformanceMetricsLogging() {
+    val workManager = WorkManager.getInstance(ApplicationProvider.getApplicationContext())
+
+    val inputData = Data.Builder().putString(
+      MetricLogSchedulingWorker.WORKER_CASE_KEY,
+      MetricLogSchedulingWorker.MEMORY_USAGE_WORKER
+    ).build()
+
+    val request: OneTimeWorkRequest = OneTimeWorkRequestBuilder<LogUploadWorker>()
+      .setInputData(inputData)
+      .build()
+
+    workManager.enqueue(request)
+    testCoroutineDispatchers.runCurrent()
+    val workInfo = workManager.getWorkInfoById(request.id)
+
+    assertThat(workInfo.get().state).isEqualTo(WorkInfo.State.SUCCEEDED)
+    assertThat(
+      fakePerformanceMetricsEventLogger
+        .getMostRecentPerformanceMetricsEvent()
+        .loggableMetric
+        .loggableMetricTypeCase
+    ).isEqualTo(OppiaMetricLog.LoggableMetric.LoggableMetricTypeCase.MEMORY_USAGE_METRIC)
+  }
+
+  @Test
+  fun testWorker_enqueueRequest_writeFails_verifyFailure() {
+    val workManager = WorkManager.getInstance(ApplicationProvider.getApplicationContext())
+
+    val inputData = Data.Builder().putString(
+      MetricLogSchedulingWorker.WORKER_CASE_KEY,
+      MetricLogSchedulingWorker.STORAGE_USAGE_WORKER
+    ).build()
+
+    val request: OneTimeWorkRequest = OneTimeWorkRequestBuilder<LogUploadWorker>()
+      .setInputData(inputData)
+      .build()
+
+    setUpPerformanceMetricsEventLoggerToFail()
+    workManager.enqueue(request)
+    testCoroutineDispatchers.runCurrent()
+    val workInfo = workManager.getWorkInfoById(request.id)
+
+    assertThat(workInfo.get().state).isEqualTo(WorkInfo.State.FAILED)
+    assertThat(fakePerformanceMetricsEventLogger.noPerformanceMetricsEventsPresent()).isTrue()
+  }
+
+  private fun setUpTestApplicationComponent() {
+    DaggerMetricLogSchedulingWorkerTest_TestApplicationComponent.builder()
+      .setApplication(ApplicationProvider.getApplicationContext())
+      .build()
+      .inject(this)
+  }
+
+  private fun setUpPerformanceMetricsEventLoggerToFail() {
+    // Simulate the log attempt itself failing during the job. Note that the reset is necessary here
+    // to remove the default stubbing for the mock so that it can properly trigger a failure.
+    Mockito.reset(mockPerformanceMetricsEventLogger)
```

Comment:
Do we need to mock this? We prefer not mocking first-party dependencies.

Reply:
umm, I'm following the existing pattern which exists in LogUploadWorkerTest -- there's a function (setUpEventLoggerToFail) that does the exact same thing. 

Also, I think it's okay to mock and setup a 'when' call to adequately throw an error because the end goal is to test whether the worker state = failed or not.

Reply:
The problem with mocking first-party objects is that it's much harder to guarantee API behavior and correctness as compared with using the real implementation. LogUploadWorkerTest ought to not do this either, so I don't think this is a case where following precedence is the right thing to do (plus, in the vast majority of cases we do use real implementations rather than mocks for precisely this reason).

Reply:
got it. I realised that I don't need to mock it in the first place as work requests achieve a fail result if no input data is attached to them.